TokuDB merge with Percona Server (BLD-287)

tplavcic opened the pull request:
**TICKET: BLD-287 - TokuDB merge with percona server**
Merge TokuDB repositories to Percona Server source
Checkout two other repositories in build jobs as they are needed also
Added PerconaFT and Percona-TokuBackup repositories as submodules
"tokutek-merge-pkg-changes" branch has one commit for build/packaging changes which are needed also.

MERGED REPOS:
https://github.com/percona/tokudb-engine.git
https://github.com/percona/tokudb-percona-server-5.6.git
https://github.com/percona/tokudb-backup-plugin.git

REPOS ADDED AS SUBMODULES:
https://github.com/percona/Percona-TokuBackup.git => plugin/tokudb_backup/Percona-TokuBackup
https://github.com/percona/PerconaFT.git => storage/tokudb/PerconaFT

**FULL OUTPUT OF GIT COMMANDS USED TO CREATE THIS MERGE BRANCH:**
*NOTICE: History seems to be preserved but there's no tags (in chat with George it was concluded we don't need them)*
*Also on the branch there's one last commit which is not from merge but to remove not needed files*
**1:**
git checkout 5.6
git checkout -b tokutek-merge
git remote add repo-tokudb-engine https://github.com/percona/tokudb-engine.git
git fetch repo-tokudb-engine
git checkout -b tokudb-engine repo-tokudb-engine/master
mkdir tokudb-engine
git mv COPYING tokudb-engine
git mv doc tokudb-engine
git mv mysql-test/suite/rpl mysql-test/suite/tokudb.rpl
git mv tools tokudb-engine
git mv scripts tokudb-engine
git mv README.md tokudb-engine
git add .
git commit -m "Merged tokudb-engine repo and created temp tokudb-engine directory"
git checkout tokutek-merge
git merge -m "Merge branch 'tokudb-engine' into tokutek-merge" tokudb-engine
git checkout --ours .gitignore
git add .gitignore
git commit -m "Merge branch 'tokudb-engine' into tokutek-merge"

**2:**
git remote add repo-tokudb-percona-server-5.6 https://github.com/percona/tokudb-percona-server-5.6.git
git fetch repo-tokudb-percona-server-5.6
git checkout -b tokudb-percona-server-5.6 repo-tokudb-percona-server-5.6/master
mkdir tokudb-percona-server-5.6
git mv mysql-test/suite/rpl mysql-test/suite/tokudb.rpl
git mv mysql-test/suite/parts mysql-test/suite/tokudb.parts
git mv *.bash tokudb-percona-server-5.6
git add .
git commit -m "Checked out tokudb-percona-server-5.6 repo into tokudb-percona-server-5.6 branch"
git checkout tokutek-merge
git merge -s theirs -m "Merge branch 'tokudb-percona-server-5.6' into tokutek-merge" tokudb-percona-server-5.6

**3:**
git remote add repo-tokudb-backup-plugin https://github.com/percona/tokudb-backup-plugin.git
git fetch repo-tokudb-backup-plugin
git checkout -b tokudb-backup-plugin repo-tokudb-backup-plugin/master
mkdir -p plugin/tokudb-backup-plugin
git mv backup plugin/tokudb-backup-plugin
git mv CMakeLists.txt plugin/tokudb-backup-plugin
git mv COPYING plugin/tokudb-backup-plugin
git mv README.md plugin/tokudb-backup-plugin
git mv README_tokudb_backup plugin/tokudb-backup-plugin
git mv tokudb_backup.cc plugin/tokudb-backup-plugin
git add .
git commit -m "Checked out tokudb-backup-plugin repo into tokudb-backup-plugin branch"
git checkout tokutek-merge
git merge -m "Merge branch 'tokudb-backup-plugin' into tokutek-merge" tokudb-backup-plugin

**TEST BUILDS:**
RELEASE BUILD (new job): http://jenkins.percona.com/view/TEST/job/percona-server-5.6-RELEASE-new/12/
PARAM BUILD (changed existing job): http://jenkins.percona.com/job/percona-server-5.6-param/933/
*The builds have been run on the "tokutek-merge-pkg-changes" branch since there's at least one change there which is needed for it to work fully*
There seemed to be some issue on debian 7 - release build only (param build) - but I think it was temporary only, so I'll rerun that to check

tplavcic commented:
Notice that there's a second PR with build/packaging changes here: https://github.com/tplavcic/percona-server/pull/1
Both need to be reviewed and merged at the same time. It was agreed when the review is done we will first merge this branch and then create a new PR for the second branch.
At that second PR there's more info provided on packaging changes.

laurynas-biveinis commented on mysql-test/README.tokudb:
```diff
@@ -0,0 +1,26 @@
+The MySQL tests for TokuDB are divided into common tests and tests specific to a
```
Completely outdated file?

laurynas-biveinis commented:
By no means a complete review:

```
- Why do we build jemalloc? We don't package it, and to run the
  tests can we rely on the system jemalloc?

- http://jenkins.percona.com/job/percona-server-5.6-param/913/BUILD_TYPE=debug,Host=ubuntu-vivid-64bit/consoleFull
 shows that TokuDB tests are not run

 http://jenkins.percona.com/job/percona-server-5.6-param/913/BUILD_TYPE=release,Host=ubuntu-vivid-64bit/consoleFull
 shows that TokuDB tests fail to run (different reason than above)
```

RoelVdP commented:
- @abychko needs to review this merge also re: jemalloc etc.
- Please test manually on Centos7 also, especially ps_tokudb_admin, see video I sent around few days ago

tplavcic commented:
@RoelVdP 
There's no jemalloc in the merge - that is part of the jenkins test job (trunk/param).
ps_tokudb_admin also currently didn't change - the most probable issue you have on centos7 is with selinux blocking LD_PRELOAD of jemalloc
It's a known issue: https://jira.percona.com/browse/BLD-255
We have some selinux test package in testing repo but they never were properly tested (they do work for this LD_PRELOAD issue since I tried that - but not sure how good they are from security point of view)
AlexeyB did check the other branch with packaging changes but since the latest discussions seems we'll have more changes or some different approach.

@laurynas-biveinis 
- regarding "mysql-test/README.tokudb" - it seems obsolete so I have removed it
- regarding jemalloc build - it was in TokuDB-MTR job before so I just moved it to trunk/param. I guess it was used initially as jemalloc has different locations in different distros perhaps and we also do package it currently into binary tarball - but if you wish I can remove that part and make it use the one provided by system?
- regarding the faulty test runs I need to make new runs since those artifacts are gone - I'll get back to you on this

laurynas-biveinis commented:
Is it possible to build jemalloc only in the job that packages it (or in the ancestor job as needed?)

tplavcic commented:
@laurynas-biveinis 
- regarding jemalloc - yes it's possible, but currently I'm working on building ha_tokudb.so with statically linked jemalloc so if it works I'll remove jemalloc from everywhere since there will be no need for it.
  Also I'll add PerconaFT and tokudb-hotbackup repos as submodules since seems we settled on that.
  Will ping you when I have updated branch and other stuff.

tplavcic commented:
@laurynas-biveinis @abychko 
I have updated the initial comment for PR.
PerconaFT and Percona-TokuBackup are now added as submodules.
trunk/param jobs are changed to see if tokudb is merged and in that case they run tests that previously TokuDB-MTR job was running.
Jemalloc is now statically linked to ha_tokudb.so - there's no dependency on jemalloc on packages and also mysqld_safe doesn't do LD_PRELOAD for that part.
ps_tokudb_admin script now has options "--enable-backup" and "--disable-backup" which add preload-hotbackup option to my.cnf
If that option is in my.cnf mysqld_safe will do LD_PRELOAD for libHotBackup.so if not it won't do it.
ps_tokudb_admin also checks for selinux status and if it's not in permissive or disabled mode it writes error (but only when doing installation of TokuBackup since it only uses LD_PRELOAD).
But for those packaging changes I'll make better description in:
https://github.com/tplavcic/percona-server/pull/1

tplavcic commented:
PR: https://github.com/tplavcic/percona-server/pull/1
has been updated also with latest change description

tplavcic commented:
Regarding the submodules, the jenkins jobs have been updated to do something as following to initialize those modules and pick-up the relevant code:

```
git submodule init
git submodule update
pushd storage/tokudb/PerconaFT
git checkout ${TOKUDB_TAG}
popd
pushd plugin/tokudb-backup-plugin/Percona-TokuBackup
git checkout ${TOKUBACKUP_TAG}
popd
```

RoelVdP commented:
Various issues added to BLD-287

laurynas-biveinis commented:
I'm OK with merging this. 

It will need follow-up work in MTR, maybe cmake etc, this is easier done on trunk

laurynas-biveinis commented:
Ah, is jemalloc build redudant now and can be removed?

tplavcic commented:
@laurynas-biveinis 
Jemalloc build has been removed from jenkins jobs and from binary tarball build.
The param build also doesn't have 900 errors but some <= 150 (wheezy issues where temporary and had 750 or some errors).
Take a look at here: http://jenkins.percona.com/job/percona-server-5.6-param/934/
(and nevermind the centos5 build error - it's because some docker testing)

Did you look at https://github.com/tplavcic/percona-server/pull/1 also?
It's mostly packaging but there are some cmake changes and stuff (jemalloc static...) - might be good to take a look.

laurynas-biveinis commented:
OK. I looked at tplavcic#1 too, looks fine with me (there are open questions but we are getting to them)

tplavcic commented:
@laurynas-biveinis @georgelorch 
New commits from tokudb-engine repo are added to this branch (nothing else has been touched).
I have created new PR for packaging changes based on the fact that we decided to not statically link with jemalloc and not to merge tokudb rpm/deb package into server package - we just have one source tarball and one binary tarball.
The other PR is here (and will be needed to make another one once this is merged): https://github.com/tplavcic/percona-server/pull/2

Param build: http://jenkins.percona.com/job/percona-server-5.6-param/953/
Test release build: http://jenkins.percona.com/view/TEST/job/percona-server-5.6-RELEASE-new/32/

laurynas-biveinis commented:
LGTM

george-lorch commented:
+1
